### PR TITLE
Viewport row buffering while scrolling

### LINKED ISSUE #402: Implement viewport row buffering while scrolling
The master branch doesn't update buffered rows if they fall outside the viewport while scrolling. The beta branch, however, updates ALL the buffered rows regardless of scrolling.

## Expected Behavior
Beta does the above optimization and saves lot of render time (usually dependent on rows within the view port).

The black window covers the total rows rendered by FDT. 
The green rows represents the total rows that lies within the viewport.
The red rows represents the lows that's outside the viewport.
fdt scroll buffer story
_(Kind of an ugly pic)_
The red rows are the problem here. In beta, these are updated regardless of scrolling, whereas Master cleverly handles their update only after the scroll is finished. This makes sense as the user won't be seeing them anyway and there's a high chance for them to disappear while scrolling is in progress.

## Current Behavior
Beta is kind of slow because of this.

## Possible Solution
While scrolling is turned on, only update the buffered rows which lie within the view port.

## Steps to Reproduce (for bugs)

In FixedDataTableBufferedRows.js, add the following POC in the render method.

```js
    const oldMap = this._staticRowArray.map((o) => {
      return [o.key, o.props.index];
    });

   // ...
   // ... existing code which updates this_staticRowArray
   // ...

    const newMap = this._staticRowArray.map((o) => {
      return [o.key, o.props.index];
    });

    oldMap.sort((a, b) => (a.props || {}).index < (b.props || {}).index);
    newMap.sort((a, b) => (a.props || {}).index < (b.props || {}).index);
    const changedMap = [];
    oldMap.forEach((x, i, a) => {
        if (oldMap[i][1] != newMap[i][1]) {
          changedMap.push([oldMap[i][1], newMap[i][1]]);
        }
      }
    );
    
    console.log('Updated buffer length = ', changedMap.length);
    console.log('Actual Buffer length = ', this._staticRowArray.length);
```

In master, you'll that only half the buffered rows got updated while scrolling. And the remaining get's updated while scrolling is finished.

## Context
Should solve lot of scroll related performance issues in beta.

## Your Environment
* Version used: v1.0.0.12-beta
* Browser Name and version: Google Chrome (lastest)
* Operating System and version (desktop or mobile): Windows 10
* Link to your project: <Our own Examples>

### PR DESCRIPTION
**Pull request by pradeepnschrodinger**
Implement viewport row buffering while scrolling.

## Description
See #402

## Motivation and Context
Fixes #402

## How Has This Been Tested?

Just manual testing done on fixed height rows.
I have used the POC in the issue for verifying that the only the rows in the viewport get updated.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: src/reducers/computeRenderedRows.js [hunk 1]
```diff
@@ -42,20 +44,25 @@ export default function computeRenderedRows(state, scrollAnchor) {
...
-  // leave only a subset of rows shown, but no scrollbar to scroll up to the first rows.
-  if (maxScrollY === 0) {
-    if (rowRange.firstViewportIdx > 0) {
-      rowRange = calculateRenderedRowRange(newState, {
-        firstOffset: 0,
-        lastIndex: rowsCount - 1,
-      });
+  // if we are scrolling, only update the offsets for the rows which are
+  // inside the viewport
+  if (state.scrolling) {
```
Comment (pradeepnschrodinger) [inline src/reducers/computeRenderedRows.js]: So this is where we branch out the logic.
Only change here is that we call `calculateViewPortRenderedRowOffsets` instead of  `calculateRenderedRowRange` if `scrolling` is true.

File: src/reducers/computeRenderedRows.js [hunk 2]
```diff
@@ -216,26 +223,103 @@ function computeRenderedRowOffsets(state, rowRange) {
...
+  // there's a possibility that the old rows (outside the viewport) overlap the new ones since scrollY was changed.
+  // To avoid this, we need to update the rowOffsets for them.
+  const unchangedRows = intersection(bufferMapping, bufferMappingOld);
+  unchangedRows.forEach(rowIdx => rowOffsetsCache[rowIdx] = rowOffsetsCacheOld[rowIdx]);
+
   state.rowHeights = rowOffsetsCache;
   state.rows = bufferMapping;
 }
+
+function addRowToBuffer(rowIdx, bufferSetStartIndex, bufferSetEndIndex, bufferSet, maxBufferSize) {
```
Comment (pradeepnschrodinger) [inline src/reducers/computeRenderedRows.js]: This function simply adds the row ID into the buffer, and returns the new index.
Used by both computeRenderedRowOffsets, and computeViewPortRenderedRowOffsets.

Comment (pradeepnschrodinger) [inline src/reducers/computeRenderedRows.js]: Change variable names, and add JSDoc

File: src/reducers/index.js [hunk 3]
```diff
@@ -144,9 +144,16 @@ function reducers(state = getInitialState(), action) {
...
+      const newState = Object.assign({}, state, {
         scrolling: false,
       });
+      const unchangedRowRange = {
+        firstIndex: state.firstRowIndex,
+        firstOffset: state.firstRowOffset,
+        lastIndex: state.lastIndex,
+        scrollJumpedY: false,
+      };
+      return computeRenderedRows(newState, unchangedRowRange);
```
Comment (pradeepnschrodinger) [inline src/reducers/index.js]: `SCROLL_END` will recompute the rows, and this time since `scrolling` is false, the entire viewport will be updated.

**Comment by wcjordan** [comment]
Could you walk me through why you chose to handle the logic in this way?  My instinct would have been to pass to the views info about which rows are in the viewport and only update those rows in the view during the render while scrolling is true.  

It seems like your current approach could lead to unnecessary mounts and unmounts when scrolling starts and ends.  Not sure how much that affects performance, but it's worth looking at.

**Comment by pradeepnschrodinger** [comment]
> Could you walk me through why you chose to handle the logic in this way? My instinct would have been to pass to the views info about which rows are in the viewport and only update those rows in the view during the render while scrolling is true.

This was my first approach too. But I had a small problem where the old rows might overlap the new rows.

During recomputing of row indexes, the rows which are removed aren't strictly within the viewport. IntegerBufferSet doesn't have the API to replace values within a specified range, but [only outside a range](https://github.com/schrodinger/fixed-data-table-2/blob/v1.0-beta/src/vendor_upstream/struct/IntegerBufferSet.js#L73). Thus it's possible that a row inside the old viewport is still present, and they get rendered at the same place, overlapping the newer ones.

The fix for this is to recompute their offsets which is [better done at the reducer level](https://github.com/schrodinger/fixed-data-table-2/pull/403/files#diff-9f78d2d8c60668be66483cb00c5c6617R297). 

> It seems like your current approach could lead to unnecessary mounts and unmounts when scrolling starts and ends. Not sure how much that affects performance, but it's worth looking at.

Only the new rows are calculated, and old rows outside the viewport are still rendered, hence there won't be any unmounts/mounts. Also the number of buffered rows is always constant.

File: src/reducers/computeRenderedRows.js [hunk 4]
```diff
@@ -216,26 +223,103 @@ function computeRenderedRowOffsets(state, rowRange) {
...
+    // Get position for the viewport row
+    const rowPosition = addRowToBuffer(rowIdx, firstViewportIdx, endViewportIdx, bufferSet, renderedRowsCount);
     bufferMapping[rowPosition] = rowIdx;
   }
 
+  // Each time we make a set of row offsets, but we only change the position for the rows inside the buffer
+  // there's a possibility that the old rows (outside the viewport) overlap the new ones since scrollY was changed.
+  // To avoid this, we need to update the rowOffsets for them.
+  const unchangedRows = intersection(bufferMapping, bufferMappingOld);
+  unchangedRows.forEach(rowIdx => rowOffsetsCache[rowIdx] = rowOffsetsCacheOld[rowIdx]);
```
Comment (pradeepnschrodinger) [inline src/reducers/computeRenderedRows.js]: If the user scrolls just 1 row, wouldn't this still update the rowOffsets for all the rows (minus the new one) ?

File: src/reducers/index.js (hunk 3 above)
Comment (pradeepnschrodinger) [reply src/reducers/index.js]: see if state.scrollJumpedY can be used instead of false

File: src/reducers/computeRenderedRows.js (hunk 4 above)
Comment (pradeepnschrodinger) [reply src/reducers/computeRenderedRows.js]: fixed with commit 68eaa38

File: src/reducers/computeRenderedRows.js (hunk 2 above)
Comment (pradeepnschrodinger) [reply src/reducers/computeRenderedRows.js]: done with 68eaa38

File: src/reducers/index.js (hunk 3 above)
Comment (pradeepnschrodinger) [reply src/reducers/index.js]: yep, it can be true/false. Doesn't affect buffering

**Comment by pradeepnschrodinger** [comment]
> @pradeepnschrodinger, check that buffering doesn't break while controlled scrolling takes place.

@pradeepnschrodinger , doesn't seem to break anything.

**Comment by wcjordan** [comment]
> I had a small problem where the old rows might overlap the new rows.
> 
> During recomputing of row indexes, the rows which are removed aren't strictly within the viewport. IntegerBufferSet doesn't have the API to replace values within a specified range, but [only outside a range](https://github.com/schrodinger/fixed-data-table-2/blob/v1.0-beta/src/vendor_upstream/struct/IntegerBufferSet.js#L73). Thus it's possible that a row inside the old viewport is still present, and they get rendered at the same place, overlapping the newer ones.

That makes sense.  But this still seems pretty complicated and requires supporting a forked logic.  I'm still digging into it a bit, so I may just be struggling to understand it.  I would have envisioned an easier way of handling this where we pass the first and last row index in the viewport to the component for rendering rows.  Then we could check for each row index if it's in the viewport.  If it's not, position it offscreen, but don't update it otherwise.

File: src/reducers/computeRenderedRows.js [hunk 5]
```diff
@@ -176,8 +178,9 @@ function calculateRenderedRowRange(state, scrollAnchor) {
 }
 
 /**
- * Walk the rows to render and compute the height offsets and
- * positions in the row buffer.
+ * Walk the rows to render that are present in the viewport and compute the height offsets and
+ * positions in the row buffer. (Also modifies the height offsets for the rows outside the viewport, to avoid
```
Comment (wcjordan) [inline src/reducers/computeRenderedRows.js]: Probably don't need to wrap this comment in parenthesis.

File: src/reducers/computeRenderedRows.js [hunk 6]
```diff
@@ -189,9 +192,10 @@ function calculateRenderedRowRange(state, scrollAnchor) {
  *   firstBufferIdx: number,
  *   firstViewportIdx: number,
  * }} rowRange
+ * @param {boolean=} computeWithinViewportOnly
  * @private
  */
-function computeRenderedRowOffsets(state, rowRange) {
+function computeRenderedRowOffsets(state, rowRange, computeWithinViewportOnly = false) {
```
Comment (wcjordan) [inline src/reducers/computeRenderedRows.js]: Do we ever call this without the 3rd parameter?  If not, let's not make it optional

File: src/reducers/computeRenderedRows.js [hunk 7]
```diff
@@ -207,35 +211,68 @@ function computeRenderedRowOffsets(state, rowRange) {
     return;
   }
 
-  const bufferMapping = []; // state.rows
+  // output for this function
+  const bufferMapping = state.rows;
```
Comment (wcjordan) [inline src/reducers/computeRenderedRows.js]: Why do we start from the old rows list here?  Is this just to ensure we stick to the same size / don't evict unless necessary?  

It seems a bit obscure of a way to accomplish that.  And I think it could end up w/ the same row rendered in 2 places (although that may not be problematic).

File: src/reducers/computeRenderedRows.js [hunk 8]
```diff
@@ -207,35 +211,68 @@ function computeRenderedRowOffsets(state, rowRange) {
...
   const rowOffsetsCache = {}; // state.rowHeights
```
Comment (wcjordan) [inline src/reducers/computeRenderedRows.js]: rowOffsetsCache is going to have endBufferIdx-firstBufferIdx items in it.  Is there any risk of that being less than the number of items in bufferMapping if we have a larger number of rows in the buffer on the previous render?  Has any work been done to avoid rows rendered in the buffer changing by +/-1 as we scroll?

File: src/reducers/computeRenderedRows.js [hunk 9]
```diff
@@ -207,35 +211,68 @@ function computeRenderedRowOffsets(state, rowRange) {
...
+    // Get position for the viewport row
+    const rowPosition = addRowToBuffer(rowIdx, bufferSet, firstViewportIdx, endViewportIdx, renderedRowsCount);
     bufferMapping[rowPosition] = rowIdx;
   }
 
+  if (computeWithinViewportOnly) {
+    // In the above loop, we only calculated row offsets for the rows inside the buffer.
+    // Now we calculate the row offsets for the remaining rows.
+    const rowsNotInBuffer = filter(state.rows, (rowIdx) => !inRange(rowIdx, firstBufferIdx, endBufferIdx));
+    rowsNotInBuffer.forEach(rowIdx => rowOffsetsCache[rowIdx] = state.rowHeights[rowIdx]);
```
Comment (wcjordan) [inline src/reducers/computeRenderedRows.js]: This seems incorrect to me.  Shouldn't the rowOffsetsCache be the full offest row positioning, not the rows height?  

In theory you could just have all these be offset off screen anywhere though as long as they're not in the viewport.  This seems a weird way to do that though.  

It does solve the concern I raise above about having different # of items in bufferMapping & rowOffsetsCache though...

File: src/reducers/computeRenderedRows.js [hunk 10]
```diff
@@ -207,35 +211,68 @@ function computeRenderedRowOffsets(state, rowRange) {
...
 
+    // Get position for the viewport row
+    const rowPosition = addRowToBuffer(rowIdx, bufferSet, firstViewportIdx, endViewportIdx, renderedRowsCount);
     bufferMapping[rowPosition] = rowIdx;
   }
 
+  if (computeWithinViewportOnly) {
+    // In the above loop, we only calculated row offsets for the rows inside the buffer.
+    // Now we calculate the row offsets for the remaining rows.
+    const rowsNotInBuffer = filter(state.rows, (rowIdx) => !inRange(rowIdx, firstBufferIdx, endBufferIdx));
```
Comment (wcjordan) [inline src/reducers/computeRenderedRows.js]: Should this be over bufferMapping instead of state.rows?  Does state.rows risk being out of date at this point?

Comment (pradeepnschrodinger) [reply src/reducers/computeRenderedRows.js]: > Should this be over bufferMapping instead of state.rows? 

I think bufferMapping and state.rows will be the same. So there's no point in assigning at the end.

However, to fix the controlled scrolling issue, we throw away the buffer (previous behavior) and use a fresh array. So this means I'll still need the assign at the end.

> Does state.rows risk being out of date at this point?

I don't think that'll happen since the reducer updates are guaranteed to be non-preemptive, if that's what you mean.
Also bufferMapping is always the same as state.rows (and on the new commit, only when computeWithinViewportOnly is true), hence this shouldn't be a problem. 
I'll still update it to bufferMapping as it leads to confusion (I just got confused while looking at this again)

File: src/reducers/computeRenderedRows.js (hunk 8 above)
Comment (pradeepnschrodinger) [reply src/reducers/computeRenderedRows.js]: > This is fixed by keeping the old rowOffsets in the rowsNotInBuffer.forEach call below.

Unfortunately, there's a bug here. This fix is only done if `scrolling` is being done. 
So to fix the fix, I defaulted back to the old behavior of having a fresh empty array of `bufferMapping` while not `scrolling`. Basically this means that the entire logic of keeping the old rows + offsets are only done while scrolling (exactly as previous behavior without buffering). 
This issue must have crept in after the refactor where I merged the 2 `computeRenderedRowOffsets` functions)

File: src/reducers/computeRenderedRows.js (hunk 9 above)
Comment (wcjordan) [reply src/reducers/computeRenderedRows.js]: Ah okay let's definitely file the ticket for renaming to rowOffsets to cleanup.  

Could you walk me through how this ensures that a row from rowsNotInBuffer won't have an offset from the previous scroll which places into the buffer (or viewport)?  

The scenario I'm concerned about might look like: 
Row X is rendered at position 200, and I'm scrolled so 250 is the end of the buffer.
I scroll up so 210 is at the end of the buffer, but in scrolling up, I find the rows newly added to the beginning of the buffer are larger than expected and Row X is actually positioned at 220.
Could Row X be out of the buffer, but positioned in the viewport since it uses the old positioning of 200 instead of the new one of 220?

File: src/reducers/computeRenderedRows.js (hunk 7 above)
Comment (wcjordan) [reply src/reducers/computeRenderedRows.js]: Could you take a pass at cleaning it up to remove the obscurity for why it's here?  It's not clear why it's needed from the way the code is written, so I fear it will easily be broken.

File: src/reducers/computeRenderedRows.js (hunk 8 above)
Comment (wcjordan) [reply src/reducers/computeRenderedRows.js]: I'm concerned about rowOffsets & the mapping being allowed to drift.  It violates certain expectations of how the code is written and may make it difficult to maintain.  Could you modify the change to keep the data structures in sync so there are reasonable invariants about how they can be used together?  

If there is a need to handle rows in the viewport and rows outside of the viewport separately, let's split those into 2 data structures to create cleaner code.  We can use a selector to merge those data structures where needed.

File: src/reducers/computeRenderedRows.js (hunk 10 above)
Comment (wcjordan) [reply src/reducers/computeRenderedRows.js]: I think this gets back to code operating unclearly over rows in viewport vs outside of viewport.  Let's split things out so there's less room for error.  

My concern is you're assuming no code above modifies any entries which are not inRange, but that may not be true for future changes to the repo.  It's better if you write code which makes less assumptions of that nature, or if you can't you should ensure those assumptions are well documented.

**Comment by pradeepnschrodinger** [comment]
> I've asked for some cleanup. There's some risk of comments / comment threads being lost within Github when pushing this many changes. Could you close this PR as abandoned and open a fresh one with the changes addressed when you're ready?
> 
> If the comments are unclear, please schedule time with me to discuss. Thanks!

@wcjordan, yeah, you're right. This PR has become hard to follow :|

I'll abandon and create a new one with the changes, but only right after #412 is fixed.
